### PR TITLE
dask: datetime `Data` init, and combination with `Query` objects

### PR DESCRIPTION
**Pull request by davidhassell**
Hi Sadie, bit rambling, but fixes a bug that didn't allow `Data` objects to be instantiated from datetime objects, and also one that prevented constructs from being combined with Query objects, in some circumstances (the latter uncovered the former).

### CODE REVIEW

**Comment by davidhassell** [comment]
... hold on, might have found a _third_ bug that can be fixed here ...

**Comment by davidhassell** [comment]
The thrid bug was actually https://github.com/NCAS-CMS/cfdm/pull/212 in `cfdm`, now merged so good to go, here (Needs the cfdm fix for the new tests to work)

**Comment by sadielbartholomew** [comment]
> now merged so good to go, here

Nice. In that case I'll set to 'Ready for review' (which somehow I get the option of doing, not just yourself as PR author!) just so I can track this without assuming it's not ready...

File: cf/data/creation.py [hunk 1]
```diff
@@ -81,22 +82,46 @@ def to_dask(array, chunks, **from_array_options):
 
     **Examples**
 
-    >>> to_dask([1, 2, 3])
+    >>> cf.data.creation.to_dask([1, 2, 3])
```
Comment (sadielbartholomew) [inline cf/data/creation.py:85]:
```suggestion
    >>> cf.data.creation.to_dask([1, 2, 3], 'auto')
```

File: cf/data/creation.py [hunk 2]
```diff
@@ -81,22 +82,46 @@ def to_dask(array, chunks, **from_array_options):
...
+    dask.array<array, shape=(1,), dtype=object, chunksize=(1,), chunktype=numpy.ndarray>
 
     """
+    if not (
+        is_dask_collection(array)
+        or isinstance(
+            array, (np.ndarray, list, tuple, memoryview) + np.ScalarType
+        )
+    ):
+        if not hasattr(array, "shape"):
```
Comment (sadielbartholomew) [inline cf/data/creation.py:103]: Is there a reason you chose not to combine the two `if` statements into one here, for foreseen extensibility perhaps?

File: cf/data/creation.py [hunk 3]
```diff
@@ -81,22 +82,46 @@ def to_dask(array, chunks, **from_array_options):
...
     kwargs.setdefault("lock", lock)
     kwargs.setdefault("meta", getattr(array, "_dask_meta", None))
-    return da.from_array(array, chunks=chunks, **kwargs)
+
+    try:
+        return da.from_array(array, chunks=chunks, **kwargs)
+    except NotImplementedError as e:
+        if (
+            str(e)
+            == "Can not use auto rechunking with object dtype. We are unable to estimate the size in bytes of object data"
```
Comment (sadielbartholomew) [inline cf/data/creation.py:121]: Maybe:

```suggestion
            == "Can not use auto rechunking with object dtype. Unable to estimate the size in bytes of object data"
```

in line with our passive 'house style' and also to be slightly more concise?

**Review by sadielbartholomew — APPROVED** [review]
I've raised some minor comments but overall this is all good, in particular (given the lack of accompanying Issue here, especially) I confirmed that the new tests are suitably thorough and that they expose the bugs in question by failing on the current `lama-to-dask` when I copied them line-for-line over there, but that they pass on this branch.

Please merge when ready after considering the minor comments. Thanks.

#### Notes

The errors I saw effectively before the changes in this PR using the new tests in this PR were, for reference:

```pytb
======================================================================
ERROR: test_Data__init__datetime (__main__.DataTest)
Test `Data.__init__` for datetime objects.
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_Data.py", line 4107, in test_Data__init__datetime
    d = cf.Data(a)
  File "/home/sadie/cf-python/cf/data/data.py", line 502, in __init__
    array = to_dask(array, chunks, **kwargs)
  File "/home/sadie/cf-python/cf/data/creation.py", line 99, in to_dask
    return da.from_array(array, chunks=chunks, **kwargs)
  File "/home/sadie/anaconda3/envs/cf-env/lib/python3.8/site-packages/dask/array/core.py", line 3353, in from_array
    chunks, x.shape, dtype=x.dtype, previous_chunks=previous_chunks
AttributeError: 'cftime._cftime.datetime' object has no attribute 'shape'
```
and:

```pytb
======================================================================
ERROR: test_Field_combine_with_Query (__main__.FieldTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "test_Field.py", line 2777, in test_Field_combine_with_Query
    (x == q).array
  File "/home/sadie/cf-python/cf/mixin/propertiesdatabounds.py", line 203, in __eq__
    return self._binary_operation(y, "__eq__", False)
  File "/home/sadie/cf-python/cf/mixin/propertiesdatabounds.py", line 504, in _binary_operation
    new = super()._binary_operation(other, method)
  File "/home/sadie/cf-python/cf/mixin/propertiesdata.py", line 599, in _binary_operation
    new.set_data(new_data, copy=False)
  File "/home/sadie/cf-python/cf/mixin/propertiesdata.py", line 5273, in set_data
    data = _Data(data, copy=False)
  File "/home/sadie/cf-python/cf/data/data.py", line 502, in __init__
    array = to_dask(array, chunks, **kwargs)
  File "/home/sadie/cf-python/cf/data/creation.py", line 99, in to_dask
    return da.from_array(array, chunks=chunks, **kwargs)
  File "/home/sadie/anaconda3/envs/cf-env/lib/python3.8/site-packages/dask/array/core.py", line 3353, in from_array
    chunks, x.shape, dtype=x.dtype, previous_chunks=previous_chunks
AttributeError: 'NotImplementedType' object has no attribute 'shape'
```

File: cf/cfdatetime.py [hunk 4]
```diff
@@ -456,20 +469,23 @@ def dt2Dt(x, calendar=None):
 
 
 def dt2rt(array, units_in, units_out, dummy1=None):
-    """Round to the nearest millisecond. This is only necessary whilst
-    netCDF4 time functions have an accuracy of no better than 1
-    millisecond (which is still the case at version 1.2.2).
+    """Return numeric time values from datetime objects.
 
-    The returned array is always independent.
+    .. seealso:: `rt2dt`
```
[the remark this answers is not in the document]
Comment (davidhassell) [reply cf/cfdatetime.py:474]: :) We need more names like that.

File: cf/data/creation.py (hunk 3 above)
Comment (davidhassell) [reply cf/data/creation.py]: I agree, but this message is produced by `dask`, and we can't control it. It's not very robust, though, having our logic controlled by the exact format and wording of someone else's error message. How about https://github.com/NCAS-CMS/cf-python/pull/441/commits/6adeaae2d93d7772b809176df2fbcdcc7cd9064f?

File: cf/data/creation.py (hunk 2 above)
Comment (davidhassell) [reply cf/data/creation.py]: No reason. Rolled into one test: 67113887cfd2866978b6ead41b788ec25e2c8e6e

File: cf/data/creation.py (hunk 3 above)
Comment (davidhassell) [reply cf/data/creation.py]: With the fix: a0a2749b0adf1faac9e982f3ab0951ec60ed8ff4! I realise that autoochunking can still occur if chunks is not `"auto"`. E.g. in ...

```python
>>> da.from_array(np.arange(200).reshape(10, 20), chunks={1: 5})
dask.array<array, shape=(10, 20), dtype=int64, chunksize=(10, 5), chunktype=numpy.ndarray>
```
... dimension `0` was auto-chunked.

Comment (sadielbartholomew) [reply cf/data/creation.py]: > I agree, but this message is produced by dask, and we can't control it.

Ah, I see. Not passive enough from Dask then! Far too friendly for my liking :smile: ...

I liked 6adeaae very much; now you suggest it, it seems much more robust to test against the chunking argument. But a0a2749 is certainly sufficient given your final note, so we can consider this resolved.

**Review by sadielbartholomew — APPROVED** [review]
All feedback addressed well, thanks. Quick final sanity check done inclusive of the new commits. Please merge.

**Pull request merged**